webapp/app/routes/app/emergency/$id/execute-emergency-upgrade-button.tsx

```diff
@@ -0,0 +1,108 @@
+import { Button } from "@/components/ui/button";
+import { ALL_ABIS } from "@/utils/raw-abis";
+import type React from "react";
+import { toast } from "react-hot-toast";
+import { type Hex, encodeAbiParameters, hexToBigInt, isAddressEqual } from "viem";
+import { useWriteContract } from "wagmi";
+
+type Signature = { signer: Hex; signature: Hex };
+type Proposal = {
+  calldata: Hex;
+  targetAddress: Hex;
+  salt: Hex;
+  value: string;
+};
+
+export type ExecuteEmergencyUpgradeButtonProps = {
+  children?: React.ReactNode;
+  boardAddress: Hex;
+  signatures: Signature[];
+  allGuardians: Hex[];
+  allCouncil: Hex[];
+  zkFoundationAddress: Hex;
+  proposal: Proposal;
+};
+
+function encodeSignatures(signatures: Signature[]): Hex {
+  return encodeAbiParameters(
+    [
+      { name: "addresses", type: "address[]" },
+      { name: "signatures", type: "bytes[]" },
+    ],
+    [signatures.map((sig) => sig.signer), signatures.map((sig) => sig.signature)]
+  );
+}
+
+export function ExecuteEmergencyUpgradeButton({
+  children,
+  boardAddress,
+  signatures,
+  allGuardians,
+  allCouncil,
+  zkFoundationAddress,
+  proposal,
+}: ExecuteEmergencyUpgradeButtonProps) {
+  const { writeContract } = useWriteContract();
+
+  const guardianSignatures = signatures
+    .filter((s) => allGuardians.some((member) => isAddressEqual(s.signer, member)))
+    .sort((a, b) => Number(hexToBigInt(a.signer) - hexToBigInt(b.signer)));
+
+  const councilSignatures = signatures
+    .filter((s) => allCouncil.some((member) => isAddressEqual(s.signer, member)))
+    .sort((a, b) => Number(hexToBigInt(a.signer) - hexToBigInt(b.signer)));
+
+  const zkFoundationSignature = signatures.find((s) =>
+    isAddressEqual(s.signer, zkFoundationAddress)
+  );
+
+  const enabled =
```

Comment:
Rather than client side logic whether to determine whether to enable button or not, what do you reckon about just reading the emergencyProposal status column if READY or not? That way DB is source of truth in a single place

Reply:
When I did this there were no states on the db, yet. Acutally, I believe they are not there yet. I can make that change once I incorporate your changes, no problem.

Reply:
Done